### PR TITLE
gh-107431: Make `multiprocessing.managers.{DictProxy,ListProxy}` generic

### LINKED ISSUE #107431: multiprocessing manager classes DictProxy and ListProxy don't support typing in 3.11.4
# Bug report
`DictProxy` and `ListProxy` classes from `multiprocessing.managers` do not support generic annotations.

Minimal reproducible example for `DictProxy`:
```py
from multiprocessing.managers import DictProxy
from multiprocessing import Manager

d: DictProxy[str, float] = Manager().dict()
```

Error:
```error
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: type 'DictProxy' is not subscriptable
```

Similar for `ListProxy`

Side note: https://github.com/python/cpython/issues/99509 appears to still be an issue as well

# Your environment
Python 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)] on win32

### Linked PRs
* gh-107433

### PR DESCRIPTION
**Pull request by sobolevn**
This looks like a reasonable use-case for generic types. 
In `typeshed` they are generic:
1. https://github.com/python/typeshed/blob/0d8a6bc2004b7b6d8335e35b017438125c305264/stdlib/multiprocessing/managers.pyi#L68
2. https://github.com/python/typeshed/blob/0d8a6bc2004b7b6d8335e35b017438125c305264/stdlib/multiprocessing/managers.pyi#L122

* Issue: gh-107431

Fixes #107431.

### CODE REVIEW

File: Lib/multiprocessing/managers.py
Comment (AlexWaygood) [inline Lib/multiprocessing/managers.py]: nit: don't think there's any reason to make a change to this line

```suggestion
    'keys', 'pop', 'popitem', 'setdefault', 'update', 'values'
```

File: Lib/multiprocessing/managers.py
Comment (AlexWaygood) [inline Lib/multiprocessing/managers.py]: I'd prefer it if we followed the style used elsewhere in this module, which would have the happy side-effect of meaning that this PR has a smaller diff :)

https://github.com/python/cpython/blob/d0dcd27d3a8b8680803510fee59836cd732feb7a/Lib/multiprocessing/managers.py#L1178-L1193

```suggestion
BaseDictProxy._method_to_typeid_ = {
    '__iter__': 'Iterator',
}


class DictProxy(BaseDictProxy):
    __class_getitem__ = classmethod(types.GenericAlias)
```

**Review by AlexWaygood — COMMENTED** [review]
Some small points:

**Review by AlexWaygood — APPROVED** [review]
Thanks, LGTM! Will wait to see if Jelle has any objections before merging.

Previous additions of `__class_getitem__` methods have been treated as features, so we probably can't backport this, unfortunately.

**Comment by AlexWaygood** [comment]
Hmm... I suppose one reason for _not_ making these subscriptable at runtime _might_ be that these are undocumented classes that aren't included in `__all__`. But I personally still lean towards making them subscriptable, nonetheless.

**Comment by JelleZijlstra** [comment]
Maybe these classes should be documented? I'm also a bit hesitant about this if these classes are undocumented.

**Comment by AlexWaygood** [comment]
> Maybe these classes should be documented? I'm also a bit hesitant about this if these classes are undocumented.

@gpshead, any thoughts from you on this? I think there's a decent case for documenting these classes, since instances of `ListProxy` and `DictProxy` are returned from `multiprocessing.managers.SyncManager.list()` and `multiprocessing.managers.SyncManager.dict()` respectively. Both of those are public, documented methods:

- https://docs.python.org/3/library/multiprocessing.html#multiprocessing.managers.SyncManager.list
- https://docs.python.org/3/library/multiprocessing.html#multiprocessing.managers.SyncManager.dict

File: Lib/multiprocessing/managers.py
Comment (vstinner) [inline Lib/multiprocessing/managers.py]: Would it be possible to make it private? I mean, just add an underscore prefix: _BaseDictProxy.

**Review by vstinner — APPROVED** [review]
LGTM, but i'm not a typing expert.

In case of doubt, I prefer to not document DictProxy and ListProxy. I don't think that not documenting these types should block this enhancement.

**Comment by AlexWaygood** [comment]
@vstinner, this needs input from multiprocessing experts, not typing experts. Two typing experts have already given their input here (Jelle and myself). From a typing perspective, the change is fine. But from a multiprocessing perspective: the classes are currently undocumented and not included in `__all__`. Adding `__class_getitem__` might encourage their use in type annotations. Is that okay? Should we document these classes, or are they best left undocumented? What's our position on that?

**Comment by vstinner** [comment]
@pitrou: Do you have an opinion on this change?

**Comment by gpshead** [comment]
I don't think anyone wants to claim expertise for multiprocessing no matter how much we've done work on it.  This change looks good to me, I made a couple edits.

**Review by gpshead — APPROVED** [review]

**Comment by pitrou** [comment]
I have no idea what a `GenericAlias` is, and the doc is rather cryptic. As long as this doesn't change runtime behavior I see no reason to oppose it.

**Comment by vstinner** [comment]
Thanks for making _BaseDictProxy private.

**Pull request merged**
